drivers: interrupt_controller: Add driver for NXP's irqsteer

LaurentiuM1234 opened the pull request:
This patch series adds a new interrupt controller driver used for NXP's irqsteer IP.
This patch series also includes the switch to using the new IRQ_STEER driver for NXP_ADSP_IMX8M SoC.

LaurentiuM1234 committed:
drivers: interrupt_controller: Add driver for NXP's IRQ_STEER IP

This commit introduces a new interrupt controller driver used
for NXP's IRQ_STEER IP.

Apart from introducing the driver itself, this commit contains
the following changes:
	1) Switch i.MX8MP to using the XTENSA core interrupt
	controller instead of the dummy irqsteer one.
		* this is required because the binding for the
		irqsteer driver is no longer a dummy one
		(since it's being used by the irqsteer driver).
		As such, to avoid having problems, switch to
		using another dummy binding.
	2) Modify the irqsteer dummy binding such that it
	serves the IRQ_STEER driver's needs.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
manifest: Bump up hal_nxp revision

Bump up hal_nxp revision to contain the IRQ_STEER-related
changes.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 commented:
Since forcing all of the i.MX8 DSP-based boards would break SOF (Sound Open Firmware) (because of the fact that XTOS is yet to be deprecated on these boards), I can't make the switch to the new IRQ_STEER driver. To better understand how the driver can be used, you may find an example of a DTS + the necessary configurations (for i.MX8MP) below:

```
soc {
    irqsteer: interrupt-controller@30a80000 {
        compatible = "nxp,irqsteer-intc";
        reg = <0x30a80000 DT_SIZE_K(64)>;
        
        #size-cells = <0>;
        #address-cells = <1>;

        master0: interrupt-controller@0 {
            compatible = "nxp,irqsteer-master";
            reg = <0>;
            interrupt-controller;
            #interrupt-cells = <1>;
            interrupts-extended = <&clic 19 0 0>;
        };

        master1: interrupt-controller@1 {
            /* some properties go here */
            interrupts-extended = <&clic 20 0 0>;
        };

          master2: interrupt-controller@1 {
            /* some properties go here */
            interrupts-extended = <&clic 21 0 0>;
        };
    };

    /* assuming our DUMMY IP uses system interrupt 34 */
    dummy_ip: my_dummy_ip@deadbeef {
        /* some more properties go here */
        interrupt-parent = <&master1>;
        interrupts = <2>;
    };
};
```

```
CONFIG_MULTI_LEVEL_INTERRUPTS=y
CONFIG_2ND_LEVEL_INTERRUPTS=y
CONFIG_2ND_LVL_ISR_TBL_OFFSET=32
CONFIG_MAX_IRQ_PER_AGGREGATOR=64
CONFIG_NUM_2ND_LEVEL_AGGREGATORS=3
CONFIG_2ND_LVL_INTR_00_OFFSET=19
CONFIG_2ND_LVL_INTR_01_OFFSET=20
CONFIG_2ND_LVL_INTR_02_OFFSET=21
```

dbaluta commented on soc/xtensa/nxp_adsp/common/soc.c:
```diff
@@ -20,6 +20,10 @@
 
+
+/* temporary workaround the fact that not all i.MX8 SoCs use IRQ_STEER driver */
+#ifndef CONFIG_NXP_IRQSTEER
```
Can't we just disable the dts node until everything is ready?

LaurentiuM1234 replied:
Depends. If i.MX8ULP uses IRQSTEER this would mean having to hold off all other i.MX8 boards (3 of them) until i.MX8ULP is ready (meaning HAL support as well), which may or may not be ok. If not, then I can just enable IRQ_STEER on i.MX8QM and QXP in this PR and solve this.

LaurentiuM1234 replied:
ACK, thanks! Will also enable IRQ_STEER for QM and QXP in this patch. In the meantime, I'll have to check if ULP still requires the definitions from `soc.c` and decide what to do with it.

LaurentiuM1234 replied:
Update: removed commit which enables the IRQ_STEER driver on i.MX8MP

iuliana-prodan commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -0,0 +1,522 @@
+	cfg = dispatcher->dev->config;
+	irq_num = IRQSTEER_GetMasterIrqCount((IRQSTEER_Type *)cfg->regmap_phys,
+					     dispatcher->master_index);
+	i = 0;
```
You can initialize `i` and `counter` on declaration, above.

LaurentiuM1234 replied:
Ah, this is more of a personal preference I guess. I want to keep the declaration separate from the initialization.

iuliana-prodan commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -0,0 +1,522 @@
+			counter++;
+		}
+
+		/* TODO: what if asserts are disabled? Do we just let the
```
Why don't you check the `counter` in an `if` and if it exceeds the threshold `break` and print an error message?

LaurentiuM1234 replied:
Mm, won't the ISR still get called anyways in that case? The point of the `ASSERT()` was to break your system whenever you get an interrupt that keeps getting asserted.

I'm thinking about changing the `ASSERT()` to a `z_irq_spurious(NULL)` call which will for sure break your system. Technically, you weren't able to fully handle your interrupt so I guess I can make a case for calling `z_irq_spurious(NULL)` for other than just breaking the system. Let me know what you think

iuliana-prodan replied:
The problem here is actually that `IRQSTEER_GetMasterNextInterrupt` always  returns the first active interrupt - this is, in fact, get first interrupt. So, if, let's say, interrupt 1 is triggered endlessly you always handle this interrupt (or, a maximum of `counter` times), never get to the next one from `CHn_STATUS`.

IMO, a proper fix will be to get all active interrupts (have a `IRQSTEER_GetMasterActiveInterrupts()`)  outside of `while()` and in `while()` handle each of these interrupts. 
In case another peripheral interrupt is trigger while you handle these, you'll have the `irqsteer_isr_dispatcher()` called again.

So, no need for a threshold.

LaurentiuM1234 replied:
The possible issue that I see with this approach is that you'll no longer be able to distinguish between the case where `irqsteer_isr_dispatcher()` is called under normal circumstances and the case where it's called when you have a "rogue" interrupt. The point of the threshold is to try to detect the anomalies and at least print an error message so that the user knows it's an interrupt-related issue. Of course, this is really dependent on the scenario (e.g: you can have an interrupt getting triggered abnormally fast for some reason but not fast enough that it will get discovered by the threshold).

I guess another potential issue with the non-`IRQSTEER_GetMasterNextInterrupt()` approach is that you're going to have to add some extra complexity in the driver since a master may have 2 groups of 32 interrupts (master1 and master2, i.MX8MP) aggregated by it instead of just 1 (master0, i.MX8MP). At this point I'm wondering if the non-`IRQSTEER_GetMasterNextInterrupt()` approach is really worth it?

iuliana-prodan replied:
> The possible issue that I see with this approach is that you'll no longer be able to distinguish between the case where `irqsteer_isr_dispatcher()` is called under normal circumstances and the case where it's called when you have a "rogue" interrupt. The point of the threshold is to try to detect the anomalies and at least print an error message so that the user knows it's an interrupt-related issue. Of course, this is really dependent on the scenario (e.g: you can have an interrupt getting triggered abnormally fast for some reason but not fast enough that it will get discovered by the threshold). **EDIT**: At this point this is just a theoretical talk on my side as I haven't really explored the case in which an interrupt gets triggered endlessly.
> 
A "rogue" interrupt should not be handle in irq_steer, this is a peripheral issue - the irq is not cleared properly or something else.

> I guess another potential issue with the non-`IRQSTEER_GetMasterNextInterrupt()` approach is that you're going to have to add some extra complexity in the driver since a master may have 2 groups of 32 interrupts (master1 and master2, i.MX8MP) aggregated by it instead of just 1 (master0, i.MX8MP). At this point I'm wondering if the non-`IRQSTEER_GetMasterNextInterrupt()` approach is really worth it?

First, `IRQSTEER_GetMasterNextInterrupt` is definitely not NextInterrupt, is FirstInterrupt.
Second, the approach of getting all active interrupt and handle them in `while()` I believe is much simpler.
You can have an uint64_t and each bit represents the active interrupt.
In case of 2 masters each with 32 interrupts, you can have first 32 bits for one master, and the rest for the other one.

iuliana-prodan replied:
In the current design I see the following problems:
1. if a peripheral has an issue and triggers interrupts you never move to the next one - you handle that one for 5 times and exit;
2. iterate over 64 (or in case of master 0, 32) interrupts for 5 times and might handle only 1 interrupt - see 1.

With the proposed solution you'll iterate only over the activate interrupts which might be max 64.

LaurentiuM1234 replied:
ACK, will fix this

dbaluta commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -0,0 +1,522 @@
+ *	into a single/multiple ones.
+ *
+ * 2) How does the IRQ_STEER IP work?
+ *	- below you may find a diagram meant to give you
```
Does Zephyr has 100 chars per line limit? I think you can expand the lines up to that limit and the comments will be easier to read.

LaurentiuM1234 replied:
Fixed, lmk if it's better.

dbaluta commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -0,0 +1,522 @@
+	if (irq_get_level(irq) == 1) {
+		/* LEVEL 1 interrupts are DSP direct */
+		z_xtensa_irq_enable(XTENSA_IRQ_NUMBER(irq));
+		return;
```
Shouldn't you also look at `enable` parameter here?

LaurentiuM1234 replied:
Yup, fixed now with V2.

dbaluta commented on drivers/interrupt_controller/intc_nxp_irqsteer.c:
```diff
@@ -0,0 +1,522 @@
+	/* find dispatcher responsible for this interrupt */
+	for (i = 0; i < ARRAY_SIZE(dispatchers); i++) {
+		if (dispatchers[i].irq == parent_irq) {
+			cfg = dispatchers[i].dev->config;
```
I would go with 

``if (dispatchers[i].irq != parent_irq)
     continue;
 ``
 
 So that you remove a level of identation and make code easier to follow

LaurentiuM1234 replied:
Thanks for the suggestion, fixed this.

LaurentiuM1234 commented:
**V2 changes**:

- various documentation improvements/fixes pointed out by @dbaluta and @iuliana-prodan.
- fixed issue with `z_soc_irq_enable_disable()` not disabling LEVEL 1 interrupts.
- ISR dispatcher now queries the status of all interrupts for its master and then iterates over the status register to find the asserted relative interrupts.
- added `UINT_TO_IRQSTEER()` macro used to cast a variable to `IRQSTEER_Type *`.
- driver now includes `sw_isr_common.h` for `z_get_sw_isr_table_idx()`'s definition.

dbaluta commented:
@LaurentiuM1234 thanks this looks good to me! We need to sort out nxp hal dependencies and we are good to merge!

dbaluta commented:
Also please sort out twister run failures:

```
 -- Found BOARD.dts: /__w/zephyr/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts
devicetree error: <Node /soc/uart@30a60000 in '/__w/zephyr/zephyr/misc/empty_file.c'> has an 'interrupts' property, but neither the node nor any of its parents has an 'interrupt-parent' property
-- In: /__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr, command: /usr/bin/python;/__w/zephyr/zephyr/scripts/dts/gen_defines.py;--dts;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/zephyr.dts.pre;--dtc-flags;'';--bindings-dirs;/__w/zephyr/zephyr/dts/bindings;--header-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/include/generated/devicetree_generated.h.new;--dts-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/zephyr.dts.new;--edt-pickle-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/edt.pickle;--edtlib-Werror;--vendor-prefixes;/__w/zephyr/zephyr/dts/bindings/vendor-prefixes.txt
CMake Error at /__w/zephyr/zephyr/cmake/modules/dts.cmake:268 (message):
  gen_defines.py failed with return code: 1
Call Stack (most recent call first):
```

https://github.com/zephyrproject-rtos/zephyr/actions/runs/6929546819/job/18847501991?pr=62776

LaurentiuM1234 commented:
> Also please sort out twister run failures:
> 
> ```
>  -- Found BOARD.dts: /__w/zephyr/zephyr/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m.dts
> devicetree error: <Node /soc/uart@30a60000 in '/__w/zephyr/zephyr/misc/empty_file.c'> has an 'interrupts' property, but neither the node nor any of its parents has an 'interrupt-parent' property
> -- In: /__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr, command: /usr/bin/python;/__w/zephyr/zephyr/scripts/dts/gen_defines.py;--dts;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/zephyr.dts.pre;--dtc-flags;'';--bindings-dirs;/__w/zephyr/zephyr/dts/bindings;--header-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/include/generated/devicetree_generated.h.new;--dts-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/zephyr.dts.new;--edt-pickle-out;/__w/zephyr/zephyr/module_tests/nxp_adsp_imx8m/sample.audio.sof/zephyr/edt.pickle;--edtlib-Werror;--vendor-prefixes;/__w/zephyr/zephyr/dts/bindings/vendor-prefixes.txt
> CMake Error at /__w/zephyr/zephyr/cmake/modules/dts.cmake:268 (message):
>   gen_defines.py failed with return code: 1
> Call Stack (most recent call first):
> ```
> 
> https://github.com/zephyrproject-rtos/zephyr/actions/runs/6929546819/job/18847501991?pr=62776

Sorry for the slip-up. This should be fixed now.

danieldegrasse commented on dts/xtensa/nxp/nxp_imx8m.dtsi:
```diff
@@ -67,15 +69,17 @@
 			/* TODO: This INTID is just a dummy
```
Should this TODO be removed? Or do we still not support UART interrupts with this PR?

iuliana-prodan replied:
Not yet!
I've explained [here](https://github.com/zephyrproject-rtos/zephyr/pull/64637#discussion_r1397363785) why.
Briefly, after this PR is merged we can enabled UART interrupt by adding the irq_steer interrupt control in the dtsi + enabling some configs. But, will still have to wait a bit in order to not break SOF.

mmahadevan108 commented on drivers/interrupt_controller/Kconfig.nxp_irqsteer:
```diff
@@ -0,0 +1,14 @@
+	default y
+	depends on DT_HAS_NXP_IRQSTEER_INTC_ENABLED
+	depends on MULTI_LEVEL_INTERRUPTS
+	depends on XTENSA
```
Do we need this, can we remove incase this is used in other NXP architectures.

mmahadevan108 commented on dts/bindings/interrupt-controller/nxp,irqsteer-master.yaml:
```diff
@@ -0,0 +1,15 @@
+description: i.MX IRQ_STEER master
+
+compatible: "nxp,irqsteer-master"
```
Where is this new binding being used.

mmahadevan108 commented on dts/xtensa/nxp/nxp_imx8m.dtsi:
```diff
@@ -17,6 +17,16 @@
+			#size-cells = <0>;
+
+			clic: interrupt-controller@0 {
+				compatible = "cdns,xtensa-core-intc";
```
Looks like no SoC is using `nxp,irqsteer-intc` compatible, can you clarify who will use the new interrupt controller driver

LaurentiuM1234 replied:
i.MX8QM, QXP and MP. We can't enable it at the moment as it would break SOF. See example above of how enabling it would look like on MP.

mmahadevan108 replied:
Sorry I don't see the example. I am trying to understand why this driver is needed if no one is using it.

LaurentiuM1234 replied:
Because it's going to become a requirement later on when we switch SOF to using Zephyr native drivers on QM, QXP, and MP. I've decided to introduce the driver right now (despite knowing we can't enable it on any of said platforms ATM) instead of later on as it has no dependencies and it would ease the review process and save us some time.

iuliana-prodan replied:
@mmahadevan108 there was also a discussion [here](https://github.com/zephyrproject-rtos/zephyr/pull/64637#discussion_r1397363785), about enabling UART interrupts on i.MX8MP, which will use irq-steer.
